### ISSUE TITLE
exception in PooledObject finalizer

### ISSUE THREAD

**Issue by kilroytrout** [pomma89/ObjectPool#10]
I had a curious exception from the PooledObject finalizer:

System.NullReferenceException: Object reference not set to an instance of an object.
   at CodeProject.ObjectPool.PooledObject.HandleReAddingToPool(Boolean reRegisterForFinalization)
   at CodeProject.ObjectPool.PooledObject.Finalize()    
   at CodeProject.ObjectPool.PooledObject.HandleReAddingToPool(Boolean reRegisterForFinalization)
   at CodeProject.ObjectPool.PooledObject.Finalize()

Somehow Finalize() entered from HandleReAddingToPool and a null exception followed. The only thing I can guess is maybe the GC.ReregisterForFinalize calls finalize again??

I wonder if you any ideas.

**Comment by kilroytrout** [comment]
To clarify the above exception is caught in my appdomain unhandled exception handler. It's preceded by an exception thrown in the factory method passed to an ObectPool:

var pool = new ObjectPool<PooledObjectWrapper<MyObject>>(() =>MyObjectCreate());

In this case MyObjectCreate throws an exception in a particular scenerio. This results in a partially initialized pooled object. Since PooledObjectWrapper descends from PooledObject, a PoolObject instance gets created on the managed heap. Some time later the finalizer (~PooledObject()) gets called by the CLR which in turn calls HandleReAddingToPool(). In my case I added debug statements to discover that in HandleReAddingToPool(), the PooledObjectInfo member is null resulting in the null ref exception being thrown back to the finalizer:

 private void HandleReAddingToPool(bool reRegisterForFinalization)
        {
            // Only when the object is reserved it can be readded to the pool.
            if (PooledObjectInfo.State == PooledObjectState.Disposed || PooledObjectInfo.State == PooledObjectState.Available)
            {
              return;
            }

Weirdly only on my .NET 4.5 Win 2012 server is PooledObjectInfo null in this case. I created a test case where the same scenerio run on my Win 10 (.NET 4.7) machine and PooledObjectInfo is not null. In any case I fixed by trapping the exception in the finalizer:

~PooledObject()
        {
           try
           {
            // Resurrecting the object
            HandleReAddingToPool(true);
            }
            catch (Exception e)
            {
            // shouldn't happen but we don't want an exception thrown from the finalizer thread
            }
        }
